Data/ByteString/Internal.hs

```diff
@@ -671,16 +671,16 @@ c2w = fromIntegral . ord
 {-# INLINE c2w #-}
 
 -- | Selects words corresponding to white-space characters in the Latin-1 range
--- ordered by frequency.
 isSpaceWord8 :: Word8 -> Bool
-isSpaceWord8 w =
-    w == 0x20 ||
-    w == 0x0A || -- LF, \n
-    w == 0x09 || -- HT, \t
-    w == 0x0C || -- FF, \f
-    w == 0x0D || -- CR, \r
-    w == 0x0B || -- VT, \v
-    w == 0xA0    -- spotted by QC..
+isSpaceWord8 w8 =
+    -- Avoid the cost of narrowing arithmetic results to Word8,
+    -- the conversion from Word8 to Word is free.
+    let w :: Word
+        !w = fromIntegral w8
+     in w - 0x21 > 0x7e   -- not [x21..0x9f]
```

Comment:
This condition discriminates 127 out of 256 possibilities. Could you please benchmark `w .&. 0x50 == 0`, which discriminates 192 values?

Reply:
With the `& 0x50` test, I get noticeably better results, which outperform also the proposed PR on all the test cases.
```
benchmarked words/lots of words
time                 221.2 μs   (220.8 μs .. 222.0 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 220.9 μs   (220.9 μs .. 221.1 μs)
std dev              338.0 ns   (159.2 ns .. 669.3 ns)

benchmarked words/one huge word
time                 18.07 μs   (17.95 μs .. 18.23 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 17.96 μs   (17.94 μs .. 18.01 μs)
std dev              104.3 ns   (68.50 ns .. 195.1 ns)

benchmarked words/paragraphs
time                 6.243 μs   (6.225 μs .. 6.267 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 6.236 μs   (6.231 μs .. 6.240 μs)
std dev              15.92 ns   (11.74 ns .. 25.17 ns)
```

Reply:
But I get even better results combining both filters:
```haskell
isSpaceWord8 :: Word8 -> Bool
isSpaceWord8 = \ !w8 ->
    -- Avoid the cost of narrowing arithmetic results to Word8,
    -- the conversion from Word8 to Word is free.
    let w :: Word
        !w = fromIntegral w8
     in w .&. 0x50 == 0   -- Quick non-whitespace filter
        && w - 0x21 > 0x7e -- Second non-whitespace filter
        && ( w == 0x20    -- SP
          || w - 0x09 < 5 -- HT, NL, VT, FF, CR
          || w == 0xa0 )  -- NBSP
```

```
benchmarked words/lots of words
time                 216.5 μs   (215.6 μs .. 218.3 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 215.8 μs   (215.7 μs .. 216.3 μs)
std dev              720.2 ns   (154.3 ns .. 1.500 μs)

benchmarked words/one huge word
time                 16.77 μs   (16.61 μs .. 16.94 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 16.87 μs   (16.82 μs .. 16.91 μs)
std dev              138.9 ns   (97.75 ns .. 183.6 ns)

benchmarked words/paragraphs
time                 6.060 μs   (6.040 μs .. 6.083 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 6.048 μs   (6.044 μs .. 6.054 μs)
std dev              16.40 ns   (12.03 ns .. 23.45 ns)
```

Just reran the PR as-is as a sanity check that nothing changed in the mean-time and I get:
```
benchmarked words/lots of words
time                 228.5 μs   (227.6 μs .. 229.0 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 229.8 μs   (229.1 μs .. 232.5 μs)
std dev              3.588 μs   (216.7 ns .. 7.228 μs)

benchmarked words/one huge word
time                 22.45 μs   (22.37 μs .. 22.52 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 22.62 μs   (22.54 μs .. 22.80 μs)
std dev              374.5 ns   (181.8 ns .. 612.1 ns)

benchmarked words/paragraphs
time                 6.746 μs   (6.727 μs .. 6.760 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 6.776 μs   (6.760 μs .. 6.841 μs)
std dev              87.14 ns   (23.18 ns .. 190.9 ns)
```

Reply:
Between the two filters only 33 candidate characters are left:
```
λ> length $ filter (\w -> w .&. 0x50 == 0 && w - 0x21 > 0x7e) [0..255 :: Word8]
33
```

Reply:
Cool stuff.

Reply:
What's more, other than whitespace, almost all are infrequent in text strings (rather than binary data):
```
[0,1,2,3,4,5,6,7,8 -- controls
,9,10,11,12,13 -- whitespace
,14,15 -- controls
,32,160 -- whitespace
,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175 -- ¡¢£¤¥¦§¨©ª«¬­®¯
]
```

Reply:
@ethercrow I see you've switch to the implementation I was testing, are you seeing similar benchmark improvements on your hardware?

Reply:
Yes, the version with two filters is the fastest for me as well.